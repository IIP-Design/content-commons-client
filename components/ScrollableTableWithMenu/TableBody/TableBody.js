import React from 'react';
import { useQuery } from '@apollo/react-hooks';
import PropTypes from 'prop-types';
import update from 'immutability-helper';
import orderBy from 'lodash/orderBy';
import { getLangTaxonomies } from 'lib/utils';
import { getProjectsType, setProjectsQueries, setProjectTitle } from 'lib/graphql/util';
import { TEAM_VIDEO_PROJECTS_QUERY } from 'lib/graphql/queries/video';
import { TEAM_PACKAGES_QUERY } from 'lib/graphql/queries/package';
import { Table } from 'semantic-ui-react';
import TableRow from 'components/ScrollableTableWithMenu/TableRow/TableRow';
import TableBodyLoading from './TableBodyLoading';
import TableBodyError from './TableBodyError';
import TableBodyNoResults from './TableBodyNoResults';
import TableBodyNoProjects from './TableBodyNoProjects';
import './TableBody.scss';

const normalizeTypesData = type => {
  const thumbnail = () => {
    if ( !type.thumbnails ) return {};

    return {
      signedUrl: type.thumbnails && type.thumbnails.length ? type.thumbnails[0].signedUrl : '',
      alt: type.thumbnails && type.thumbnails.length ? type.thumbnails[0].alt : ''
    };
  };

  const normalizedTypeData = Object.create( {}, {
    __typename: { value: type.__typename },
    id: { value: type.id },
    createdAt: { value: type.createdAt },
    updatedAt: { value: type.updatedAt },
    projectTitle: { value: setProjectTitle( type ) },
    author: { value: `${type.author ? type.author.firstName : ''} ${type.author ? type.author.lastName : ''}` },
    team: { value: type.team ? type.team.name : '' },
    status: { value: type.status || '' },
    visibility: { value: type.visibility },
    thumbnail: { value: thumbnail() },
    categories: { value: getLangTaxonomies( type.categories ) }
  } );

  return normalizedTypeData;
};

const normalizeDashboardData = types => {
  const normalizedProjects = [];

  types.forEach( type => normalizedProjects.push( normalizeTypesData( type ) ) );

  return normalizedProjects;
};

const TableBody = props => {
  const {
    bodyPaginationVars,
    column,
    direction,
    searchTerm,
    selectedItems,
    tableHeaders,
    toggleItemSelection,
    variables,
    projectTab,
    team
  } = props;

  // Determine type of dashboard projects
  const dashboardProjectsType = getProjectsType( team );

  // Determine which Query to run
  const graphQuery = setProjectsQueries( team, {
    videoProjects: TEAM_VIDEO_PROJECTS_QUERY,
    packages: TEAM_PACKAGES_QUERY
  } );

  // Run Query
  const {
    loading, error, data
  } = useQuery( graphQuery, {
    variables: column === 'author' ? { ...variables } : { ...variables, ...bodyPaginationVars },
    notifyOnNetworkStatusChange: true,
    fetchPolicy: 'cache-and-network'
  } );

  if ( error ) return <TableBodyError error={ error } />;
  // Checks for existing data so loading doesn't flash when resort columns
  if ( loading && !data ) return <TableBodyLoading />;

  // Set dashboard projects data & ID's for status subscription
  const setDashboardProjectsData = () => {
    let projectsData;

    if ( dashboardProjectsType === 'videoProjects' ) projectsData = data.videoProjects;
    if ( dashboardProjectsType === 'packages' ) projectsData = data.packages;

    return projectsData || null;
  };

  const dashboardProjects = setDashboardProjectsData();

  if ( !dashboardProjects ) return null;
  if ( searchTerm && !dashboardProjects.length ) return <TableBodyNoResults searchTerm={ searchTerm } />;
  if ( !dashboardProjects.length ) return <TableBodyNoProjects />;

  const normalized = normalizeDashboardData( dashboardProjects );

  // No option exisiting in schema to order by author
  // When author column is clicked, a query is sent to the server without pagination variables
  // This results in an overfetch that has to be filtered on the the client side
  const order = direction ? `${direction === 'ascending' ? 'asc' : 'desc'}` : 'desc';

  const authorSorting = orderBy(
    normalized,
    tableDatum => {
      const formattedTableDatum = tableDatum.author;

      return formattedTableDatum;
    },
    [order]
  );

  // skip & first query vars are used as start/end slice() params to paginate tableData on client
  const { skip, first } = bodyPaginationVars;
  const paginatedAuthorSorting = authorSorting.slice( skip, skip + first );

  const tableData = column === 'author' ? paginatedAuthorSorting : normalized;

  return (
    <Table.Body className="projects">
      { tableData.map( d => (
        <TableRow
          key={ d.id }
          d={ d }
          selectedItems={ selectedItems }
          tableHeaders={ tableHeaders }
          toggleItemSelection={ toggleItemSelection }
          projectTab={ projectTab }
          team={ team }
        />
      ) ) }
    </Table.Body>
  );
};

TableBody.propTypes = {
  bodyPaginationVars: PropTypes.object,
  column: PropTypes.string,
  searchTerm: PropTypes.string,
  selectedItems: PropTypes.object,
  tableHeaders: PropTypes.array,
  toggleItemSelection: PropTypes.func,
  variables: PropTypes.object,
  direction: PropTypes.string,
  projectTab: PropTypes.string,
  team: PropTypes.object
};

export default TableBody;

