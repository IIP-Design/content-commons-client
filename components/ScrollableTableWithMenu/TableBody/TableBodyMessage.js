import React, { Fragment } from 'react';
import PropTypes from 'prop-types';
import { Loader, Table } from 'semantic-ui-react';

import ApolloError from 'components/errors/ApolloError';

const setMessage = ( type, error, searchTerm ) => {
  switch ( type ) {
  case 'error' && error:
    return <ApolloError error={ error } />;
  case 'loading':
    return (
      <Fragment>
        <Loader active inline size="small" />
        <span style={ { marginLeft: '0.5em', fontSize: '1.5em' } }>
          Loading...
        </span>
      </Fragment>
    );
  case 'no-projects':
    return 'No projects';
  case 'no-results' && searchTerm:
    return `No results for &ldquo;${searchTerm}&rdquo;`;
  default:
    return '';
  }
};

const TableBodyMessage = ( { error, searchTerm, type } ) => (
  <Table.Body>
    <Table.Row>
      <Table.Cell>
        {setMessage( type, error, searchTerm ) }
      </Table.Cell>
    </Table.Row>
  </Table.Body>
);

TableBodyMessage.propTypes = {
  error: PropTypes.object,
  searchTerm: PropTypes.string,
  type: PropTypes.string
};

export default TableBodyMessage;
