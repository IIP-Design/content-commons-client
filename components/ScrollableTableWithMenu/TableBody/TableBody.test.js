import { mount } from 'enzyme';
import toJSON from 'enzyme-to-json';

import TableBody from './TableBody';

/**
 * Use custom element to avoid "incorrect casing" error msg
 * @see https://jestjs.io/docs/en/tutorial-react.html#snapshot-testing-with-mocks-enzyme-and-react-16
 */
jest.mock( 'next/dynamic', () => () => 'my-project-primary-col' );

const props = {
  bodyPaginationVars: {
    first: 2,
    orderBy: 'createdAt_DESC',
    skip: 0
  },
  column: 'createdAt',
  data: [
    {
      author: 'Marek',
      categories: '',
      createdAt: '2020-04-22T13:22:08.848Z',
      id: 'ck9bd9hsh06bc07882icvx8e0',
      projectTitle: 'Guidance Package 04-22-20',
      status: 'PUBLISHED',
      team: 'GPA Press Office',
      thumbnail: {},
      updatedAt: '2020-04-22T13:22:42.672Z',
      visibility: 'INTERNAL',
      __typename: 'Package'
    },
    {
      author: 'Marek',
      categories: '',
      createdAt: '2020-04-20T16:30:48.298Z',
      id: 'ck98p4elp02i20788l3lxq5gt',
      projectTitle: 'Guidance Package 04-20-20',
      status: 'PUBLISHED',
      team: 'GPA Press Office',
      thumbnail: {},
      updatedAt: '2020-04-20T16:31:14.936Z',
      visibility: 'INTERNAL',
      __typename: 'Package'
    }
  ],
  direction: 'descending',
  error: null,
  loading: false,
  projectTab: 'teamProjects',
  searchTerm: '',
  selectedItems: new Map(),
  tableHeaders: [
    { name: 'projectTitle', label: 'PROJECT TITLE' },
    { name: 'visibility', label: 'VISIBILITY' },
    { name: 'updatedAt', label: 'MODIFIED' },
    { name: 'author', label: 'AUTHOR' }
  ],
  toggleItemSelection: jest.fn(),
  team: { contentTypes: ['VIDEO'] }
};

// Wrap TableBody in a table component to avoid warning message about invalid nesting of a <tbody>
const createTable = tableProps => (
  <table>
    <TableBody { ...tableProps } />
  </table>
);

describe( '<TableBody />', () => {
  it( 'renders initial loading state without crashing', () => {
    const newProps = { ...props, data: undefined, loading: true };

    const wrapper = mount( createTable( newProps ) );

    const message = wrapper.find( 'TableBodyMessage' );

    expect( message.exists() ).toEqual( true );
    expect( message.prop( 'type' ) ).toEqual( 'loading' );
  } );

  it( 'renders error message if an error is returned', () => {
    const newProps = { ...props, error: { message: 'There was an error.' } };

    const wrapper = mount( createTable( newProps ) );

    const message = wrapper.find( 'TableBodyMessage' );

    expect( message.exists() ).toEqual( true );
    expect( message.prop( 'type' ) ).toEqual( 'error' );
    // expect( errorComponent.contains( 'There was an error.' ) )
    // .toEqual( true );
  } );

  it( 'renders null if videoProjects or packages are null', () => {
    const newProps = { ...props, data: null };

    const wrapper = mount( createTable( newProps ) );

    const tableBody = wrapper.find( TableBody );

    expect( tableBody.html() ).toEqual( null );
  } );

  it( 'renders a no projects message if there are no projects', () => {
    const newProps = { ...props, data: [] };
    const wrapper = mount( createTable( newProps ) );

    const message = wrapper.find( 'TableBodyMessage' );

    expect( message.exists() ).toEqual( true );
    expect( message.prop( 'type' ) ).toEqual( 'no-projects' );
  } );

  it( 'renders a no results message if there are no search results', () => {
    const newProps = { ...props, data: [], searchTerm: 'new term' };

    const wrapper = mount( createTable( newProps ) );

    const message = wrapper.find( 'TableBodyMessage' );

    expect( message.exists() ).toEqual( true );
    expect( message.prop( 'type' ) ).toEqual( 'no-results' );
  } );

  it( 'renders the correct table row(s)', async() => {
    const wrapper = mount( createTable( props ) );

    const tableRows = wrapper.find( 'tr' );
    const rowCount = tableRows.length;

    const expectedRowCount = props.data.length;

    expect( rowCount ).toEqual( expectedRowCount );
    expect( toJSON( tableRows ) ).toMatchSnapshot();
  } );
} );
