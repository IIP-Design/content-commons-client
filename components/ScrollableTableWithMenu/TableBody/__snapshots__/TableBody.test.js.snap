// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TableBody /> renders the correct table row(s) 1`] = `
Array [
  <tr
    className=""
  >
    <TableCell
      as="td"
      className="items_table_item"
      data-header="PROJECT TITLE"
      key="ck9bd9hsh06bc07882icvx8e0_projectTitle"
    >
      <td
        className="items_table_item"
        data-header="PROJECT TITLE"
      >
        <div
          className="primary_col"
        >
          <my-project-primary-col
            d={
              Object {
                "__typename": "Package",
                "author": "Marek",
                "categories": "",
                "createdAt": "2020-04-22T13:22:08.848Z",
                "id": "ck9bd9hsh06bc07882icvx8e0",
                "projectTitle": "Guidance Package 04-22-20",
                "status": "PUBLISHED",
                "team": "GPA Press Office",
                "thumbnail": Object {},
                "updatedAt": "2020-04-22T13:22:42.672Z",
                "visibility": "INTERNAL",
              }
            }
            header={
              Object {
                "label": "PROJECT TITLE",
                "name": "projectTitle",
              }
            }
          />
        </div>
        <TableMobileDataToggleIcon
          isOpen={false}
          toggleDisplay={[Function]}
        >
          <Icon
            as="i"
            className="items_table_mobileDataToggleIcon"
            name="chevron down"
            onClick={[Function]}
          >
            <i
              aria-hidden="true"
              className="chevron down icon items_table_mobileDataToggleIcon"
              onClick={[Function]}
            />
          </Icon>
        </TableMobileDataToggleIcon>
      </td>
    </TableCell>
    <TableCell
      as="td"
      className="items_table_item"
      data-header="VISIBILITY"
      key="ck9bd9hsh06bc07882icvx8e0_visibility"
    >
      <td
        className="items_table_item"
        data-header="VISIBILITY"
      >
        <span>
          <div
            className="items_table_mobileHeader"
          >
            VISIBILITY
          </div>
          INTERNAL
        </span>
        <br />
      </td>
    </TableCell>
    <TableCell
      as="td"
      className="items_table_item"
      data-header="MODIFIED"
      key="ck9bd9hsh06bc07882icvx8e0_updatedAt"
    >
      <td
        className="items_table_item"
        data-header="MODIFIED"
      >
        <span>
          <div
            className="items_table_mobileHeader"
          >
            MODIFIED
          </div>
          April 22, 2020
        </span>
        <br />
      </td>
    </TableCell>
    <TableCell
      as="td"
      className="items_table_item"
      data-header="AUTHOR"
      key="ck9bd9hsh06bc07882icvx8e0_author"
    >
      <td
        className="items_table_item"
        data-header="AUTHOR"
      >
        <span>
          <div
            className="items_table_mobileHeader"
          >
            AUTHOR
          </div>
          Marek
        </span>
        <br />
      </td>
    </TableCell>
  </tr>,
  <tr
    className=""
  >
    <TableCell
      as="td"
      className="items_table_item"
      data-header="PROJECT TITLE"
      key="ck98p4elp02i20788l3lxq5gt_projectTitle"
    >
      <td
        className="items_table_item"
        data-header="PROJECT TITLE"
      >
        <div
          className="primary_col"
        >
          <my-project-primary-col
            d={
              Object {
                "__typename": "Package",
                "author": "Marek",
                "categories": "",
                "createdAt": "2020-04-20T16:30:48.298Z",
                "id": "ck98p4elp02i20788l3lxq5gt",
                "projectTitle": "Guidance Package 04-20-20",
                "status": "PUBLISHED",
                "team": "GPA Press Office",
                "thumbnail": Object {},
                "updatedAt": "2020-04-20T16:31:14.936Z",
                "visibility": "INTERNAL",
              }
            }
            header={
              Object {
                "label": "PROJECT TITLE",
                "name": "projectTitle",
              }
            }
          />
        </div>
        <TableMobileDataToggleIcon
          isOpen={false}
          toggleDisplay={[Function]}
        >
          <Icon
            as="i"
            className="items_table_mobileDataToggleIcon"
            name="chevron down"
            onClick={[Function]}
          >
            <i
              aria-hidden="true"
              className="chevron down icon items_table_mobileDataToggleIcon"
              onClick={[Function]}
            />
          </Icon>
        </TableMobileDataToggleIcon>
      </td>
    </TableCell>
    <TableCell
      as="td"
      className="items_table_item"
      data-header="VISIBILITY"
      key="ck98p4elp02i20788l3lxq5gt_visibility"
    >
      <td
        className="items_table_item"
        data-header="VISIBILITY"
      >
        <span>
          <div
            className="items_table_mobileHeader"
          >
            VISIBILITY
          </div>
          INTERNAL
        </span>
        <br />
      </td>
    </TableCell>
    <TableCell
      as="td"
      className="items_table_item"
      data-header="MODIFIED"
      key="ck98p4elp02i20788l3lxq5gt_updatedAt"
    >
      <td
        className="items_table_item"
        data-header="MODIFIED"
      >
        <span>
          <div
            className="items_table_mobileHeader"
          >
            MODIFIED
          </div>
          April 20, 2020
        </span>
        <br />
      </td>
    </TableCell>
    <TableCell
      as="td"
      className="items_table_item"
      data-header="AUTHOR"
      key="ck98p4elp02i20788l3lxq5gt_author"
    >
      <td
        className="items_table_item"
        data-header="AUTHOR"
      >
        <span>
          <div
            className="items_table_mobileHeader"
          >
            AUTHOR
          </div>
          Marek
        </span>
        <br />
      </td>
    </TableCell>
  </tr>,
]
`;
